File: apollo-integration/src/test/java/com/apollographql/apollo/ApolloIdlingResourceTest.java
package com.apollographql.apollo;


import com.google.common.truth.Truth;

import android.support.test.espresso.IdlingResource;

import com.apollographql.android.impl.normalizer.EpisodeHeroName;
import com.apollographql.android.impl.normalizer.type.Episode;
import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.espresso.ApolloIdlingResource;
import com.apollographql.apollo.exception.ApolloException;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.io.IOException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicInteger;

import javax.annotation.Nonnull;

import okhttp3.OkHttpClient;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;

import static com.google.common.truth.Truth.assertThat;

public class ApolloIdlingResourceTest {

  private ApolloIdlingResource idlingResource;
  private ApolloClient apolloClient;
  private MockWebServer server;
  private OkHttpClient okHttpClient;

  private static final String FILE_EPISODE_HERO_NAME_WITH_ID = "EpisodeHeroNameResponseWithId.json";
  private static final String IDLING_RESOURCE_NAME = "apolloIdlingResource";

  private static final long TIME_OUT_SECONDS = 3;

  @Before
  public void setup() {
    server = new MockWebServer();
    okHttpClient = new OkHttpClient.Builder()
        .build();
  }

  @After
  public void tearDown() {
    idlingResource = null;
    try {
      server.shutdown();
    } catch (IOException ignored) {

    }
  }

  @Test
  public void onNullNamePassed_NullPointerExceptionIsThrown() {

    apolloClient = ApolloClient.builder()
        .okHttpClient(okHttpClient)
        .serverUrl(server.url("/"))
        .build();

    try {
      idlingResource = ApolloIdlingResource.create(null, apolloClient);
    } catch (Exception e) {
      assertThat(e).isInstanceOf(NullPointerException.class);
      assertThat(e.getMessage()).isEqualTo("name == null");
    }
  }

  @Test
  public void onNullApolloClientPassed_NullPointerExceptionIsThrown() {
    try {
      idlingResource = ApolloIdlingResource.create(IDLING_RESOURCE_NAME, null);
    } catch (Exception e) {
      assertThat(e).isInstanceOf(NullPointerException.class);
      assertThat(e.getMessage()).isEqualTo("apolloClient == null");
    }
  }

  @Test
  public void checkValidIdlingResourceNameIsRegistered() {

    apolloClient = ApolloClient.builder()
        .okHttpClient(okHttpClient)
        .serverUrl(server.url("/"))
        .build();

    idlingResource = ApolloIdlingResource.create(IDLING_RESOURCE_NAME, apolloClient);

    Truth.assertThat(idlingResource.getName()).isEqualTo(IDLING_RESOURCE_NAME);
  }

  @Test
  public void checkIsIdleNow_whenCallIsQueued() throws IOException, TimeoutException, InterruptedException {
    server.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID));

    final NamedCountDownLatch firstLatch = new NamedCountDownLatch("firstLatch", 1);
    final NamedCountDownLatch secondLatch = new NamedCountDownLatch("secondLatch", 1);

    ExecutorService executorService = Executors.newFixedThreadPool(1);

    EpisodeHeroName query = EpisodeHeroName.builder().episode(Episode.EMPIRE).build();

    apolloClient = ApolloClient.builder()
        .okHttpClient(okHttpClient)
        .dispatcher(executorService)
        .serverUrl(server.url("/"))
        .build();

    idlingResource = ApolloIdlingResource.create(IDLING_RESOURCE_NAME, apolloClient);
    assertThat(idlingResource.isIdleNow()).isTrue();

    apolloClient.newCall(query).enqueue(new ApolloCall.Callback<EpisodeHeroName.Data>() {
      @Override public void onResponse(@Nonnull Response<EpisodeHeroName.Data> response) {
        firstLatch.countDown();
        try {
          secondLatch.await(TIME_OUT_SECONDS, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
          throw new AssertionError(e);
        }
      }

      @Override public void onFailure(@Nonnull ApolloException e) {
        throw new AssertionError("This callback can't be called.");
      }
    });
    assertThat(idlingResource.isIdleNow()).isFalse();

    firstLatch.await(TIME_OUT_SECONDS, TimeUnit.SECONDS);

    secondLatch.countDown();

    executorService.shutdown();
    executorService.awaitTermination(TIME_OUT_SECONDS, TimeUnit.SECONDS);
    assertThat(idlingResource.isIdleNow()).isTrue();
  }

  @Test
  public void checkIdlingResourceTransition_whenCallIsQueued() throws IOException, ApolloException {
    server.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID));

    apolloClient = ApolloClient.builder()
        .okHttpClient(okHttpClient)
        .dispatcher(Utils.immediateExecutorService())
        .serverUrl(server.url("/"))
        .build();

    EpisodeHeroName query = EpisodeHeroName.builder().episode(Episode.EMPIRE).build();
    final AtomicInteger counter = new AtomicInteger(1);
    idlingResource = ApolloIdlingResource.create(IDLING_RESOURCE_NAME, apolloClient);

    idlingResource.registerIdleTransitionCallback(new IdlingResource.ResourceCallback() {
      @Override public void onTransitionToIdle() {
        counter.decrementAndGet();
      }
    });

    assertThat(counter.get()).isEqualTo(1);
    apolloClient.newCall(query).execute();
    assertThat(counter.get()).isEqualTo(0);
  }

  private MockResponse mockResponse(String fileName) throws IOException {
    return new MockResponse().setChunkedBody(Utils.readFileToString(getClass(), "/" + fileName), 32);
  }
}

File: apollo-idling-resource/src/test/java/com/apollographql/apollo3/test/espresso/ApolloIdlingResourceTest.kt
package com.apollographql.apollo3.test.espresso

import androidx.test.espresso.IdlingResource.ResourceCallback
import com.apollographql.apollo3.ApolloCall
import com.apollographql.apollo3.ApolloClient
import com.apollographql.apollo3.api.Operation
import com.apollographql.apollo3.api.Query
import com.apollographql.apollo3.api.ApolloResponse
import com.apollographql.apollo3.api.ResponseField
import com.apollographql.apollo3.api.ResponseAdapterCache
import com.apollographql.apollo3.api.json.JsonReader
import com.apollographql.apollo3.api.json.JsonWriter
import com.apollographql.apollo3.api.internal.OperationRequestBodyComposer.compose
import com.apollographql.apollo3.api.ResponseAdapter
import com.apollographql.apollo3.exception.ApolloException
import com.apollographql.apollo3.rx2.Rx2Apollo
import com.google.common.truth.Truth
import okhttp3.OkHttpClient
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import okio.BufferedSource
import okio.ByteString
import org.junit.After
import org.junit.Assert
import org.junit.Rule
import org.junit.Test
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import com.apollographql.apollo3.test.espresso.ApolloIdlingResource
import org.junit.Before

class ApolloIdlingResourceTest {
  @get:Rule
  val server = MockWebServer()

  private lateinit var apolloClient: ApolloClient

  @Before
  fun setup() {
    apolloClient = ApolloClient.builder()
        .okHttpClient(OkHttpClient())
        .serverUrl(server.url("/"))
        .build()
  }

  @Test
  fun checkValidIdlingResourceNameIsRegistered() {
    val idlingResource = ApolloIdlingResource.create(IDLING_RESOURCE_NAME, apolloClient)
    Truth.assertThat(idlingResource.getName()).isEqualTo(IDLING_RESOURCE_NAME)
  }

  @Test
  @Throws(InterruptedException::class)
  fun checkIsIdleNow_whenCallIsQueued() {
    server.enqueue(mockResponse().setHeadersDelay(500, java.util.concurrent.TimeUnit.MILLISECONDS))
    val latch = CountDownLatch(1)
    val executorService = Executors.newFixedThreadPool(1)
    apolloClient = ApolloClient.builder()
        .okHttpClient(OkHttpClient())
        .dispatcher(executorService)
        .serverUrl(server.url("/"))
        .build()
    val idlingResource = ApolloIdlingResource.create(IDLING_RESOURCE_NAME, apolloClient)
    Truth.assertThat(idlingResource.isIdleNow()).isTrue()
    apolloClient.query(EMPTY_QUERY).enqueue(object : ApolloCall.Callback<Query.Data>() {
      override fun onResponse(response: ApolloResponse<Query.Data>) {
        latch.countDown()
      }

      override fun onFailure(e: ApolloException) {
        throw AssertionError("This callback can't be called.")
      }
    })
    Truth.assertThat(idlingResource.isIdleNow()).isFalse()
    latch.await(TIME_OUT_SECONDS, TimeUnit.SECONDS)
    executorService.shutdown()
    executorService.awaitTermination(TIME_OUT_SECONDS, TimeUnit.SECONDS)
    Thread.sleep(100)
    Truth.assertThat(idlingResource.isIdleNow()).isTrue()
  }

  @Test
  @Throws(InterruptedException::class)
  fun checkIsIdleNow_whenCallIsWatched() {
    server.enqueue(mockResponse())
    val latch = CountDownLatch(1)
    val executorService = Executors.newFixedThreadPool(1)

    val idlingResource = ApolloIdlingResource.create(IDLING_RESOURCE_NAME, apolloClient)
    Truth.assertThat(idlingResource.isIdleNow()).isTrue()
    apolloClient.query(EMPTY_QUERY).watcher().enqueueAndWatch(object : ApolloCall.Callback<Query.Data>() {
      override fun onResponse(response: ApolloResponse<Query.Data>) {
        latch.countDown()
      }

      override fun onFailure(e: ApolloException) {
        throw AssertionError("This callback can't be called.")
      }
    })
    Truth.assertThat(idlingResource.isIdleNow()).isFalse()
    latch.await(TIME_OUT_SECONDS, TimeUnit.SECONDS)
    executorService.shutdown()
    executorService.awaitTermination(TIME_OUT_SECONDS, TimeUnit.SECONDS)
    Thread.sleep(100)
    Truth.assertThat(idlingResource.isIdleNow()).isTrue()
  }

  @Test
  @Throws(ApolloException::class)
  fun checkIdlingResourceTransition_whenCallIsQueued() {
    server.enqueue(mockResponse())
    apolloClient = ApolloClient.builder()
        .okHttpClient(OkHttpClient())
        .dispatcher { command -> command.run() }
        .serverUrl(server.url("/"))
        .build()
    val counter = AtomicInteger(1)
    val idlingResource = ApolloIdlingResource.create(IDLING_RESOURCE_NAME, apolloClient)
    idlingResource.registerIdleTransitionCallback(ResourceCallback { counter.decrementAndGet() })
    Truth.assertThat(counter.get()).isEqualTo(1)
    Rx2Apollo.from(apolloClient.query(EMPTY_QUERY)).test().awaitTerminalEvent()
    Truth.assertThat(counter.get()).isEqualTo(0)
  }

  private fun mockResponse(): MockResponse {
    return MockResponse().setResponseCode(200).setBody("{" +
        "  \"errors\": [" +
        "    {" +
        "      \"message\": \"Cannot query field \\\"names\\\" on type \\\"Species\\\".\"," +
        "      \"locations\": [" +
        "        {" +
        "          \"line\": 3," +
        "          \"column\": 5" +
        "        }" +
        "      ]" +
        "    }" +
        "  ]" +
        "}")
  }

  companion object {
    private const val TIME_OUT_SECONDS: Long = 3
    private const val IDLING_RESOURCE_NAME = "apolloIdlingResource"
    private val EMPTY_QUERY: Query<Query.Data> = object : Query<Query.Data> {
      override fun queryDocument(): String {
        return ""
      }

      override fun serializeVariables(writer: JsonWriter, responseAdapterCache: ResponseAdapterCache) {
        writer.beginObject()
        writer.endObject()
      }

      override fun adapter(responseAdapterCache: ResponseAdapterCache): ResponseAdapter<Query.Data> {
        return object: ResponseAdapter<Query.Data> {
          override fun fromResponse(reader: JsonReader): Query.Data {
            while (reader.selectName(emptyList()) != -1) {
              // consume the json stream
            }
            return object: Query.Data {}
          }

          override fun toResponse(writer: JsonWriter, value: Query.Data) {
            TODO("Not yet implemented")
          }
        }
      }

      override fun name(): String = "EmptyQuery"

      override fun operationId(): String {
        return ""
      }
      override fun responseFields(): List<ResponseField.FieldSet> {
        return emptyList()
      }
    }
  }
}
